File: server/game.py
from server.card import Deck, Card
from server.field import Field
from server.player import Player


class Game:
    def __init__(self):
        self.players = []

    def add_player(self, name):
        self.players.append(Player(name))

        if len(self.players) == 2:
            self.start()

    def start(self):
        self.deck = Deck()
        for i in range(52):
            suit, value = divmod(i, 13)
            self.deck.add(Card(suit, value))

        self.deck.shuffle()

        for _ in range(4):
            for player in self.players:
                player.draw(self.deck)

        self.players[1].draw(self.deck)  # Second player gets extra card

        self.discard = Deck()

        self.field = Field(self.players, self.discard, self.deck, self)

        self.effects = []
        self.oneoff = None

        self.turn = 0

    def restart(self):
        self.players = [Player(player.name) for player in self.players[::-1]]

        self.start()

    def play(self, player, index, type="point", targets=[]):
        if type == "point":
            card = player.play(index)
            self.field.play_point(player, card)
        elif type == "cuttle":
            target_id = targets[0]["id"]
            opponent = self.get_opponent(player)
            card = player.hand[index]
            if card.id >= target_id:
                player.play(index)
                self.field.fields[opponent.name]["point"].remove([target_id])
                return False
            return True
        else:
            card = player.hand[index]
            opponent = self.get_opponent(player)
            if card.value in [1, 8, 10] and self.field.has_queen(opponent):
                if targets[0]["value"] != 12:
                    return True
            card = player.play(index)
            if card.value in [0, 1, 2, 3, 4, 5, 6, 8]:
                self.effects.append(card)
                self.targets = targets
                return True
            self.field.play_special(player, card, targets)

        return False

    def draw(self, player):
        player.draw(self.deck)

    def get_player(self, username):
        player = next(
            filter(
                lambda player: player.name == username,
                self.players
            ), None
        )
        return player

    def counter(self, player, index):
        card = player.play(index)

        if card.value != 1:
            player.add_card(card)
        else:
            self.effects.append(card)

    def move(self, player, move):
        if self.is_over():
            return False, None

        if player != self.players[(self.turn + len(self.effects)) % 2]:
            return False, None

        action = move["action"]
        print('ACTION', action)
        if action == "play":
            action_required = self.play(
                player,
                move["index"],
                move["type"],
                move["targets"]
            )

            if action_required:
                if move["type"] == "cuttle":
                    return False, None
                if len(self.effects) > 0:
                    self.oneoff = {
                        "effects": self.effects,
                        "turn": self.turn + len(self.effects),
                        "counter": False
                    }
                    return True, self.oneoff
                return False, None
        elif action == "draw":
            self.draw(player)
        elif action == "counter":
            self.counter(player, move["index"])
            self.oneoff = {
                "effects": self.effects,
                "turn": self.turn + len(self.effects),
                "counter": True
            }
            return True, self.oneoff
        elif action == "accept":
            if len(self.effects) % 2 != 0:
                card = self.effects[0]
                opponent = self.get_opponent(player)
                self.field.play_special(opponent, card, self.targets)

            self.effects = []
            self.oneoff = None

        self.turn += 1
        return True, None

    def get_state(self):
        return {
            "deck": len(self.deck),
            "turn": self.turn,
            "field": self.field,
            "discard": self.discard,
            "over": self.is_over(),
            "winner": self.field.get_winner()
        }

    def get_opponent(self, player):
        opponent = next(
            filter(
                lambda other: other != player,
                self.players
            )
        )
        return opponent

    def get_hands(self, player):
        if player:
            other = self.get_opponent(player)
            if self.field.can_see(player):
                other_hand = other.hand
            else:
                other_hand = [Card(-1, -1)] * len(other.hand)
            hands = {
                player.name: player.hand,
                other.name: other_hand
            }
        else:
            p1 = self.players[0]
            p2 = self.players[1]
            hands = {
                p1.name: [Card(-1, -1)] * len(p1.hand),
                p2.name: [Card(-1, -1)] * len(p2.hand)
            }

        return {
            "hands": hands
        }

    def get_players(self, player):
        if player:
            other = self.get_opponent(player)

            players = {
                "self": player.name,
                "other": other.name,
                "turn": self.players.index(player)
            }
        else:
            players = {
                "self": self.players[0].name,
                "other": self.players[1].name,
                "turn": -1
            }

        return {
            "players": players
        }

    def data(self):
        return {
            "turn": self.turn,
            "deck": self.deck,
            "discard": self.discard,
            "field": self.field
        }

    def get_data(self):
        return self.deck

    def is_over(self):
        return self.field.get_winner() is not None

File: src/Game.js
import React, { useState, useEffect } from 'react';
import io from 'socket.io-client';
import {
  EuiButton,
  EuiPage,
  EuiFlexGroup,
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiPageContent,
  EuiModalFooter,
  EuiOverlayMask,
} from '@elastic/eui';

import './App.css';

import CardList from './components/CardList';
import Card from './components/Card';
import Field from './components/Field';
import Fish from './components/Fish';

import { hasOneoff, hasPermanent, hasPoint, canCuttle, hasTarget, getTargets } from './utils/card';

const socket = io();

function Game(props) {
  const { player } = props;

  const [data, setData] = useState();
  const [players, setPlayers] = useState();
  const [hands, setHands] = useState();
  const [oneoff, setOneoff] = useState();
  const [selected, setSelected] = useState(null);
  const [target, setTarget] = useState(false);
  const [targetMode, setTargetMode] = useState('field');
  const canPlay = data?.turn % 2 === players?.turn && !data?.over;
  const [targets, setTargets] = useState([]);
  const [type, setType] = useState();
  const selectedCard = selected !== null ? hands[players.self][selected] : null;

  const updateHands = (data) => {
    setData(data);
    socket.emit('get_hands', (data) => {
      console.log('hands', data);
      setHands(data.hands);
    });
  };

  const updateData = () => {
    socket.emit('get_data', (data) => {
      console.log('data', data);
      setData(data);
    });
  };

  const updateOneoff = () => {
    socket.emit('get_oneoff', (data) => {
      console.log('oneoff', data);
      setOneoff(data);
    });
  };

  const updatePlayers = () => {
    socket.emit('get_players', (data) => {
      console.log('players', data);
      setPlayers(data.players);
    });
  };

  const selectCard = (card) => {
    setSelected(hands[players.self].indexOf(card));
  };

  const targetCard = (card) => {
    if (targets.indexOf(card) === -1) {
      setTargets((prev) => [...prev, card]);
    } else {
      setTargets((prev) => prev.filter((prev) => prev !== card));
    }
  };

  useEffect(() => {
    fetch(`/login/${player}`);

    updateData();
    updatePlayers();
    updateHands();

    socket.on('update', (data, players) => {
      console.log('updated data', data);
      if (players) {
        setData(data);
        setPlayers(players.players);
      } else {
        updateHands(data);
      }
    });

    socket.on('connect', () => {
      updateData();
      updatePlayers();
      updateHands();
      updateOneoff();
      setTarget(0);
      setSelected(null);
    });

    socket.on('restart', () => {
      updatePlayers();
      setOneoff();
      setTargets();
    });

    socket.on('players', (players) => {
      console.log('updated players', players);
      setPlayers(players.players);
    });

    socket.on('oneoff', (data) => {
      setOneoff(data);
      console.log('oneoff', data);
    });

    return () => {
      socket.removeAllListeners();
    };
  }, [player]);

  return (
    <EuiPageContent>
      <EuiFlexGroup direction="column">
        {/* <Fish /> */}
        {selected !== null ? (
          <EuiOverlayMask onClick={() => setSelected(null)}>
            <EuiModal onClose={() => setSelected(null)}>
              <EuiModalHeader>
                <EuiModalHeaderTitle>Choose an action</EuiModalHeaderTitle>
              </EuiModalHeader>

              <EuiModalBody>
                <Card card={hands[players.self][selected]} />
              </EuiModalBody>

              <EuiModalFooter>
                <EuiButton
                  disabled={!hasPoint(selectedCard)}
                  onClick={() => {
                    socket.emit(
                      'move',
                      {
                        action: 'play',
                        index: selected,
                        type: 'point',
                        targets: [],
                      },
                      () => {
                        setSelected(null);
                      },
                    );
                  }}
                  fill
                >
                  Play Value
                </EuiButton>

                <EuiButton
                  disabled={!canCuttle(selectedCard, data?.field[players.other].point.cards)}
                  onClick={() => {
                    setTarget(1);
                    setTargets([]);
                    setType('cuttle');
                    setTargetMode(['opponent', 'point']);
                  }}
                  fill
                >
                  Cuttle
                </EuiButton>

                <EuiButton
                  disabled={!hasOneoff(selectedCard) && !hasPermanent(selectedCard)}
                  onClick={() => {
                    if (!hasTarget(selectedCard)) {
                      socket.emit(
                        'move',
                        {
                          action: 'play',
                          index: selected,
                          type: 'special',
                          targets: [],
                        },
                        () => {
                          setSelected(null);
                        },
                      );
                    } else {
                      const values = getTargets(selectedCard);
                      console.log(values);

                      setTarget(values.count);
                      setTargets([]);
                      setTargetMode(values.mode);
                      setType('special');
                    }
                  }}
                  fill
                >
                  Play Effect
                </EuiButton>
              </EuiModalFooter>
            </EuiModal>
          </EuiOverlayMask>
        ) : null}
        {oneoff?.turn % 2 == players?.turn ? (
          <EuiOverlayMask>
            <EuiModal maxWidth={false}>
              <EuiModalHeader>
                <EuiModalHeaderTitle>
                  {players.other}
                  {oneoff.counter ? ' countered' : ' used an one-off card'}
                </EuiModalHeaderTitle>
              </EuiModalHeader>

              <EuiModalBody>
                <Card card={oneoff.effects[oneoff.effects.length - 1]} />
              </EuiModalBody>

              <EuiModalFooter>
                <EuiButton
                  color="danger"
                  onClick={() => {
                    socket.emit('move', {
                      action: 'accept',
                    });
                    setOneoff(null);
                  }}
                  fill
                >
                  Accept
                </EuiButton>

                <EuiButton
                  disabled={hands[players.self].filter((card) => card.value == 1).length == 0}
                  onClick={() => {
                    const card = hands[players.self].find((card) => card.value == 1);
                    const index = hands[players.self].indexOf(card);
                    socket.emit('move', {
                      action: 'counter',
                      index,
                    });
                    setOneoff(null);
                  }}
                  fill
                >
                  Counter
                </EuiButton>
              </EuiModalFooter>
            </EuiModal>
          </EuiOverlayMask>
        ) : null}
        {target ? (
          <EuiOverlayMask onClick={() => setTarget(0)}>
            <EuiModal onClose={() => setTarget(0)}>
              <EuiModalHeader>
                <EuiModalHeaderTitle>Pick a card to target</EuiModalHeaderTitle>
              </EuiModalHeader>

              <EuiModalBody>
                {data && players ? (
                  <Field
                    data={data}
                    players={players}
                    onClick={targetCard}
                    targets={targets}
                    filter={targetMode}
                  />
                ) : null}
              </EuiModalBody>

              <EuiModalFooter>
                <EuiButton
                  disabled={target != targets.length}
                  onClick={() => {
                    socket.emit(
                      'move',
                      {
                        action: 'play',
                        index: selected,
                        type,
                        targets,
                      },
                      () => {
                        setTarget(0);
                        setTargets([]);
                        setSelected(null);
                      },
                    );
                  }}
                  fill
                >
                  Confirm
                </EuiButton>
              </EuiModalFooter>
            </EuiModal>
          </EuiOverlayMask>
        ) : null}
        {data?.over ? <div>Winner: {data ? data.winner : null}</div> : null}
        Other hand
        <div className="hand">
          {hands && players ? <CardList cards={hands[players.other]} /> : null}
        </div>
        {data && players ? <Field data={data} players={players} /> : null}
        Hand
        <div className="hand">
          {hands && players ? (
            <CardList cards={hands[players.self]} onClick={canPlay ? selectCard : null} />
          ) : null}
        </div>
        {data?.over ? (
          <EuiButton
            fill
            disabled={data && data.deck === 0}
            onClick={() => {
              socket.emit('restart');
            }}
          >
            Play again
          </EuiButton>
        ) : (
          <EuiButton
            disabled={data?.deck === 0 || !canPlay}
            onClick={() => {
              socket.emit('move', {
                action: 'draw',
              });
            }}
            fill
          >
            Draw ({data?.deck ?? 0})
          </EuiButton>
        )}
      </EuiFlexGroup>
    </EuiPageContent>
  );
}

export default Game;
